### DISCUSSION TITLE
Demo Edu Registry - Institute affiliation attestation

### DISCUSSION THREAD

**Discussion by pkranga** [Sunbird-RC/community#299] [Issues]
I have set up the demo education registry schema locally and using the UI to register and manage the education use cases. When creating an institute it is asking for the affiliation details But when the affiliation details are filled it is going for attestation. As per the schema the attestation would be done by the below

"name": "instituteAffiliation",
        "conditions": "(ATTESTOR#$.[*]#.contains('board-cbse'))",
        "type": "MANUAL",
        "attestorPlugin": "did:internal:ClaimPluginActor?entity=board-cbse",
        "attestationProperties": {
          "affiliation": "$.affiliation",
          "instituteName": "$.instituteName"

How to identify which user can approve this? Should we create default master users who can onboard institutes?

**Comment by pkranga** [comment]
Without the approval we are not able to onboard teachers and students to that Institute

**Reply by tejash-jl** [reply]
@pkranga yes in this flow you need to create a user in keycloak to have a role `board-cbse`. You can use this token to get the claims that are raised and approve the attestations.

**Comment by pkranga** [comment]
I have an user with that role and still he is not able to see the request for approval in the UI. http://localhost:4200/registry/api/v1//Institute/claims API returns 0 records

The decoded jwt token is below

{
  "exp": 1670317702,
  "iat": 1670303303,
  "auth_time": 1670303302,
  "jti": "dd0b0518-5613-493c-827b-31e65b6a7a81",
  "iss": "http://keycloak:8080/auth/realms/sunbird-rc",
  "aud": "account",
  "sub": "5e2e616a-e0ae-4b43-8b47-e47ac0144b33",
  "typ": "Bearer",
  "azp": "registry-frontend",
  "nonce": "538a4916-c35c-4a24-8543-5355a2962686",
  "session_state": "27092835-fb1b-48a3-9f2f-c708e097587f",
  "acr": "1",
  "allowed-origins": [
    "https://localhost:4202",
    "http://localhost:4202",
    "https://localhost:4200",
    "https://ndear.xiv.in",
    "http://localhost:4200/*",
    "*",
    "http://localhost:4200",
    "http://ndear.xiv.in",
    "http://20.198.64.128"
  ],
  "realm_access": {
    "roles": [
      "offline_access",
      "board-cbse",
      "uma_authorization",
      "default-roles-ndear"
    ]
  },
  "resource_access": {
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
  "scope": "openid profile email",
  "email_verified": false,
  "preferred_username": "public@yopmail.com",
  "entity": [
    "Institute"
  ],
  "email": "public@yopmail.com"
}

**Reply by tejash-jl** [reply]
@pkranga i think you have added the role to the existing user in keycloak. 
Basically the `entity` in `attributes` needs to be changed. It should contain `board-cbse`
URL : {host}/admin/master/console/#/realms/sunbird-rc/users/{user-id}/user-attributes

**Reply by pkranga** [reply]
If this has to be an entity then we should have created a new entity type rt? But we have only Institute, Student and Teacher.

I update the attributes but we are getting an error that "Fetching claims failed Entity is not part of our system" when invoking this API - http://localhost:4200/registry/api/v1//board-cbse/claims

**Reply by tejash-jl** [reply]
@pkranga in this flow, we are showcasing how an external entity(who is not part of sbrc) can attest a claim.
You would need to add `external_entities` to registry
https://github.com/Sunbird-RC/demo-education-registry/blob/main/docker-compose.yml#L44

**Reply by tejash-jl** [reply]
also check if this flag is also enabled.
https://github.com/Sunbird-RC/demo-education-registry/blob/main/docker-compose.yml#L52

**Reply by pkranga** [reply]
The api will still be http://localhost:4200/registry/api/v1//board-cbse/claims or http://localhost:4200/registry/api/v1//Institute/claims. Both the API's for the user with user attribute board-cbse returns 0 records


{
  "exp": 1670329276,
  "iat": 1670314876,
  "jti": "9155cbc5-afa8-4ad7-8208-74ec21d4be42",
  "iss": "http://keycloak:8080/auth/realms/sunbird-rc",
  "aud": "account",
  "sub": "0294b077-a804-4ff5-95b1-949de4f5ff63",
  "typ": "Bearer",
  "azp": "registry-frontend",
  "session_state": "7d6d96cc-78dd-4fbc-877f-9abb250de53f",
  "acr": "1",
  "allowed-origins": [
    "https://localhost:4202",
    "http://localhost:4202",
    "https://localhost:4200",
    "https://ndear.xiv.in",
    "http://localhost:4200/*",
    "*",
    "http://localhost:4200",
    "http://ndear.xiv.in",
    "http://20.198.64.128"
  ],
  "realm_access": {
    "roles": [
      "offline_access",
      "board-cbse",
      "uma_authorization",
      "default-roles-ndear"
    ]
  },
  "resource_access": {
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
  "scope": "profile email",
  "email_verified": true,
  "name": "Rangabashyam CBSE",
  "preferred_username": "ranga.pk",
  "given_name": "Rangabashyam",
  "family_name": "CBSE",
  "entity": [
    "board_cbse"
  ],
  "email": "ranga.pk@yopmail.com"
}

Which table should i check if the affiliation request for attestation has been generated or not?

**Reply by tejash-jl** [reply]
you can check `claims` table.
Also the complete flow for education is documented here,
https://github.com/Sunbird-RC/demo-education-registry/blob/main/edu-attestation-flows.ipynb

**Reply by sajeshkayyath** [reply]
We are getting an API failure while executing the play book script (demo-education-registry/edu-attestation-flows.ipynb)

All the steps / API calls before "**cbse board attest claims**" is working fine, but this one is resulting in error as below.

`---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[102], line 12
      6 headers = {
      7   'Content-Type': 'application/json',
      8     'Authorization': 'bearer %s'%boardToken,
      9 }
     11 response = requests.request("GET", baseUrl + "board-cbse/claims", headers=headers)
---> 12 assert response.status_code == 200
     13 print(response.json())

AssertionError: `

**In the registry service logs, the below error is appearing**

`09:15:08.549 [http-nio-8081-exec-8] INFO  d.s.registry.sink.shard.ShardManager:43 - Activated shard shard1 for attribute value null
09:15:08.550 [ForkJoinPool.commonPool-worker-5] INFO  d.s.elastic.ElasticServiceImpl:121 - connection info: index:schema connectioninfo:es:9200
09:15:08.550 [ForkJoinPool.commonPool-worker-5] INFO  d.s.elastic.ElasticServiceImpl:125 - resthighclient obj:org.elasticsearch.client.RestHighLevelClient@410382cb
09:15:08.559 [http-nio-8081-exec-8] INFO  d.s.r.s.impl.RegistryServiceImpl:132 - Heath Check :  
09:15:10.514 [http-nio-8081-exec-10] ERROR d.s.r.c.RegistryClaimsController:54 - Fetching claims failed Entity is not part of our system
java.lang.Exception: Entity is not part of our system
	at dev.sunbirdrc.registry.helper.RegistryHelper.getRequestedUserDetails(RegistryHelper.java:756)
`

We verified through keycloak and confirmed that the boad-cbse role and cbse-admin user exist in the system.

 The registry version where this is being executed is as below. 

`dockerhub/sunbird-rc-core                       v0.0.13`

Please advice how to go about this issue to run the claim locally. Also let me know if any additional info is needed.

**Reply by pkranga** [reply]
@tejash-jl

**Reply by Nijinsha** [reply]
Why am I getting notifications on this?

**Reply by pkranga** [reply]
@Nijinsha You have subscribed to this topic. Please click "Unsubscribe" button to remove from the notification

**Reply by tejash-jl** [reply]
@pkranga can you check if https://github.com/Sunbird-RC/demo-education-registry/blob/main/docker-compose.yml#L44 is configured

**Reply by sajeshkayyath** [reply]
thanks @tejash-jl , adding this has fixed the issue
`external_entities=board-cbse`
